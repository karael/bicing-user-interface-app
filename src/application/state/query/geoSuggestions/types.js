import {
  FAILURE, PENDING, START, SUCCESS,
} from 'application/state/stateConstants';
import { defineAction } from 'redux-define';

export const FETCH = defineAction(
  'FETCH',
  [START, PENDING, SUCCESS, FAILURE],
  defineAction('query/geoSuggestions'),
);

export default FETCH;
