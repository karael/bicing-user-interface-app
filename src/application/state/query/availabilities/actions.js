import { FETCH } from 'application/state/query/availabilities/types';

export const fetchStart = (
  itineraryStep,
  itineraryAt,
  periodStartAt,
  periodEndAt,
  interval,
  ...stationIds
) => ({
  error: false,
  meta: { isFetching: true, itineraryStep, itineraryAt },
  payload: {
    periodStartAt, periodEndAt, interval, stationIds,
  },
  type: FETCH.START,
});
export const fetchPending = itineraryStep => ({
  error: false,
  meta: { isFetching: true, itineraryStep },
  type: FETCH.PENDING,
});
export const fetchSuccess = (itineraryStep, data) => ({
  error: false,
  meta: { isFetching: false, itineraryStep },
  type: FETCH.SUCCESS,
  payload: data,
});
export const fetchFailure = (itineraryStep, error) => ({
  error: true,
  meta: { isFetching: false, itineraryStep },
  type: FETCH.FAILURE,
  payload: error,
});
