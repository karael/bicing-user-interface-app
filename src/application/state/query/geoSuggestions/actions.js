import { FETCH } from 'application/state/query/geoSuggestions/types';

export const fetchStart = () => ({
  error: false,
  meta: { isFetching: false },
  type: FETCH.START,
});
export const fetchPending = () => ({
  error: false,
  meta: { isFetching: true },
  type: FETCH.PENDING,
});
export const fetchSuccess = geoSuggestions => ({
  error: false,
  meta: { isFetching: false },
  type: FETCH.SUCCESS,
  payload: geoSuggestions,
});
export const fetchFailure = error => ({
  error: true,
  meta: { isFetching: false },
  type: FETCH.FAILURE,
  payload: error,
});
