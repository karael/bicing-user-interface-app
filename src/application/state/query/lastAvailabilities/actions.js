import * as Types from 'application/state/query/lastAvailabilities/types';

export const fetchListStart = () => ({ type: Types.FETCH_LIST.START, payload: {} });
export const fetchListPending = () => ({ type: Types.FETCH_LIST.PENDING, payload: {} });
export const fetchListCancelled = () => ({ type: Types.FETCH_LIST.CANCELLED, payload: {} });
export const fetchListSuccess = data => ({
    type: Types.FETCH_LIST.SUCCESS,
    payload: {
        data
    },
});
export const fetchListFailure = error => ({
    type: Types.FETCH_LIST.FAILURE,
    payload: {
        error,
    },
});
