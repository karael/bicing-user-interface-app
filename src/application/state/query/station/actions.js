import * as Types from 'application/state/query/station/types';

export const fetchStart = (stationId) => ({ type: Types.FETCH.START, payload: { stationId } });
export const fetchPending = () => ({ type: Types.FETCH.PENDING, payload: {} });
export const fetchCancelled = () => ({ type: Types.FETCH.CANCELLED, payload: {} });
export const fetchSuccess = data => ({ type: Types.FETCH.SUCCESS, payload: { data } });
export const fetchFailure = error => ({ type: Types.FETCH.FAILURE, payload: { error } });
