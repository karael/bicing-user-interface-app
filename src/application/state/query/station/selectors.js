const selector = state => state.query.station;

export const data = state => selector(state).data;
export const error = state => selector(state).error;
export const isFetching = state => selector(state).isFetching;
