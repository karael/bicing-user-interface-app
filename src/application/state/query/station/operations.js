import { fetchFailure, fetchPending, fetchSuccess } from 'application/state/query/station/actions';
import { FETCH } from 'application/state/query/station/types';
import HttpStationQuery from 'infrastructure/bicingApi/HttpStationQuery';
import Joi from 'joi';
import { call, put, takeLatest } from 'redux-saga/effects';
import { stationType, stationIdType } from 'domain/types/stationType';

export function* fetch(action) {
  yield put(fetchPending());

  try {
    const { stationId } = action.payload;

    Joi.assert(stationId, stationIdType);

    const station = yield call(HttpStationQuery.find, stationId);

    Joi.assert(station, stationType);

    yield put(fetchSuccess(station));
  } catch (exception) {
    yield put(fetchFailure(exception));
  }
}

export default function* operation() {
  yield takeLatest(FETCH.START, fetch);
}
