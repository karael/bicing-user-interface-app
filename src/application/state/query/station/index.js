import * as actions from 'application/state/query/station/actions';
import operation from 'application/state/query/station/operations';
import reducer from 'application/state/query/station/reducers';
import * as selectors from 'application/state/query/station/selectors';
import * as types from 'application/state/query/station/types';

export {
  actions,
  operation,
  selectors,
  types,
};

export default reducer;
