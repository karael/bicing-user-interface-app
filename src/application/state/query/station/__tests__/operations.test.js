import * as actions from 'application/state/query/station/actions';
import operation, { fetch } from 'application/state/query/station/operations';
import { FETCH } from 'application/state/query/station/types';
import HttpStationQuery from 'infrastructure/bicingApi/HttpStationQuery';
import { expectSaga, testSaga } from 'redux-saga-test-plan';
import * as matchers from 'redux-saga-test-plan/matchers';
import { throwError } from 'redux-saga-test-plan/providers';
import { v4 as uuid } from 'uuid';

describe('application/state/query/station/operations', () => {
  test('should wait for a fetch start event to list with operation()', () => {
    testSaga(operation)
      .next()
      .takeLatestEffect(FETCH.START, fetch);
  });

  test('should dispatch a fetchPending action with fetch() generator', () => {
    testSaga(fetch)
      .next()
      .put(actions.fetchPending());
  });

  test('should list expected station with fetch() generator', () => {
    const stationId = uuid();
    const fakeStation = {
      '@id': `api/stations/${stationId}`,
      '@type': 'stationView',
      id: `${stationId}`,
      name: '233 - C/NOU DE LA RAMBLA, 164',
      type: 'BIKE',
      address: 'Nou de la Rambla',
      addressNumber: '164',
      zipCode: '08004',
      latitude: 41.371965,
      longitude: 2.166871,
    };
    const action = { type: FETCH.START, payload: { stationId } };

    return expectSaga(fetch, action)
      .provide([
        [matchers.call.fn(HttpStationQuery.find, stationId), fakeStation],
      ])
      .put(actions.fetchSuccess(fakeStation))
      .run();
  });

  test('should handle error when api call failed in fetch() generator', () => {
    const stationId = uuid();
    const error = new Error('error_api_call');
    const action = { type: FETCH.START, payload: { stationId } };

    return expectSaga(fetch, action)
      .provide([
        [matchers.call.fn(HttpStationQuery.find, stationId), throwError(error)],
      ])
      .put(actions.fetchFailure(error))
      .run();
  });
});
