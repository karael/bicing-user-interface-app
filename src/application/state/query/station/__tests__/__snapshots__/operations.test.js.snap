// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application/state/query/station/operations should handle error when api call response does not contains expected schema type 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "a41070e4-3866-490d-a8ab-5195eae71f93",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/find",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": false,
          "meta": Object {
            "isFetching": true,
          },
          "type": "query/station/FETCH_PENDING",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": true,
          "meta": Object {
            "isFetching": false,
          },
          "payload": Object {
            "_object": Object {
              "name": "233 - C/NOU DE LA RAMBLA, 164",
            },
            "annotate": "@@redux-saga-test-plan/json/function/<anonymous>",
            "details": Array [
              Object {
                "context": Object {
                  "key": "id",
                  "label": "id",
                },
                "message": "\\"id\\" is required",
                "path": Array [
                  "id",
                ],
                "type": "any.required",
              },
            ],
            "isJoi": true,
            "name": "ValidationError",
          },
          "type": "query/station/FETCH_FAILURE",
        },
        "channel": null,
      },
    },
  ],
}
`;

exports[`application/state/query/station/operations should handle error when stationId is not validated 1`] = `
Object {
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": false,
          "meta": Object {
            "isFetching": true,
          },
          "type": "query/station/FETCH_PENDING",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": true,
          "meta": Object {
            "isFetching": false,
          },
          "payload": Object {
            "_object": "not an uuid",
            "annotate": "@@redux-saga-test-plan/json/function/<anonymous>",
            "details": Array [
              Object {
                "context": Object {
                  "key": undefined,
                  "label": "value",
                  "value": "not an uuid",
                },
                "message": "\\"value\\" must be a valid GUID",
                "path": Array [],
                "type": "string.guid",
              },
            ],
            "isJoi": true,
            "name": "ValidationError",
          },
          "type": "query/station/FETCH_FAILURE",
        },
        "channel": null,
      },
    },
  ],
}
`;
